rfd/0156-notifications.md

```diff
@@ -0,0 +1,423 @@
+to be merged.
+
+We include the `auth connector` in the key to prevent the edge case where a SSO user has the same username as a local user (or a user using another auth connector), this would cause the users' notifications to be merged.
+
+Global notifications will be stored in the database under the key `/notifications/global/<notification id>`.
+
+There will be a hard cap of 10,000 notifications which can exist at any given time. If a new notification is added beyond this point, it will delete and replace the oldest existing notification.
+
+#### Types and Protos
+
+###### Notification Item
+
+If this is a user-specific notification, it will be stored as-is under `/notification/user/<cluster id>/<auth connector>/<username>/<notification id>`.
+
+If this is a global notification, it will be nested in a [`GlobalNotification`](#global-notifications).
+
+```protobuf
+// Notification represents a notification item.
+message Notification {
+   // id is the UUID of the notification, in UUIDv7 format.
+   string id = 1;
+   // code is the unique notification code.
+   // This code represents the type of notification this is, eg. `access-request-approved`.
+   string code = 2;
+   // title is the title/header of the notification.
+   string title = 3;
+   // description is the text content of the notification.
+   string description = 4;
+   // created is when the notification was created, in UNIX time.
+   google.protobuf.Timestamp created = 5;
+   // ttl is how long the notification should last before automatically expiring. Default is 30 days.
+   int64 ttl = 6;
+   // labels is optional additional metadata to be added to the notification in the form of labels.
+   map<string, string> labels = 7;
+}
+```
+
+###### User Notification State Item
+
+Each user will have a unique `NotificationState` item for every notification they have interacted with.
+
+This will keep track of whether the user has viewed (clicked on) or dismissed the notification.
+
+These will be stored per user per notification in the database under `/notification/states/<cluster id>/<auth connector>/<username>/<notification id>`
+
+Dismissed states are only going to be possible for global notifications. If a user dismisses a user-specific notification directed at them, instead of updating this state to `dismissed`, the notification will just be
+deleted as well as its state item.
+
+```protobuf
+// UserNotificationState represents a notification's state for a user. This is to keep track
+// of whether the user has viewed or dismissed the notification.
+message UserNotificationState {
+   // notification_state is the state of this notification for the user. This can be either "viewed" or "dismissed".
+   string notification_state = 1;
+}
+```
+
+##### User Last Seen Notification Item
+
+This keeps track of the timestamp of the latest notification the user has seen in their notifications list. This is used in order to determine whether
```

Comment:
I'm not clear on why we need this. If there's not a notification state of `viewed`, doesn't that mean that the notification is unread?

Reply:
This is because a `viewed` notification is different to a `read` notification. Notifications being read/unread is only determined by whether the user has opened the notifications list and seen it in the list. We only want to mark a notification as `viewed` once the user has actually clicked on it or interacted with it in some way (ie via a quick action).

Notifications that haven't been viewed (clicked on) will be highlighted in the list to differentiate them from viewed ones.

Reply:
Do I have this right? Possible states are:

- "unread" (default state)
- read (user has seen it in the list but not interacted with it)
- viewed (user has interacted with it)
- dismissed (user has dismissed the notification and doesn't want to see it any more)

If so, I'd suggest choosing different names for `read` and `viewed`. They both sound the same. In maybe we change `viewed` to `clicked`?

But also, why do we care about the difference between viewed and clicked? Do we need to render the notification differently after it's been clicked?

Reply:
That's right, I agree with the naming suggested since it can be confusing (`read` and `viewed` are essentially synonyms).

> Do we need to render the notification differently after it's been clicked?

Yes, notifications that haven't been clicked on will be highlighted in the notifications list to differentiate them from ones that have been already.